src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3757,46 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
                 .Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__Equals));
             if (isObjectEqualsMethodOrOverride ||
-                isWellKnownEqualityMethodOrImplementation(compilation, method, WellKnownMember.System_IEquatable_T__Equals))
+                isWellKnownEqualityMethodOrImplementation(compilation, method, receiverType.Type, WellKnownMember.System_IEquatable_T__Equals))
             {
                 Debug.Assert(arguments.Length == 1);
                 learnFromEqualsMethodArguments(node.ReceiverOpt, receiverType, arguments[0], results[0].RValueType);
                 return;
             }
 
-            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, WellKnownMember wellKnownMember)
+            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, TypeSymbol receiverType, WellKnownMember wellKnownMember)
             {
-                var wellKnownMethod = compilation.GetWellKnownTypeMember(wellKnownMember);
-                if (wellKnownMethod is null)
+                var wellKnownMethod = (MethodSymbol)compilation.GetWellKnownTypeMember(wellKnownMember);
+                if (wellKnownMethod is null || receiverType is null)
                 {
                     return false;
                 }
 
                 var wellKnownType = wellKnownMethod.ContainingType;
                 var parameterType = method.Parameters[0].TypeWithAnnotations;
                 var constructedType = wellKnownType.Construct(ImmutableArray.Create(parameterType));
-
-                Symbol constructedMethod = null;
-                foreach (var member in constructedType.GetMembers(WellKnownMemberNames.ObjectEquals))
-                {
-                    if (member.OriginalDefinition.Equals(wellKnownMethod))
-                    {
-                        constructedMethod = member;
-                        break;
-                    }
-                }
-
-                Debug.Assert(constructedMethod != null, "the original definition is present but the constructed method isn't present");
+                var constructedMethod = wellKnownMethod.AsMember(constructedType);
 
                 // FindImplementationForInterfaceMember doesn't check if this method is itself the interface method we're looking for
                 if (constructedMethod.Equals(method))
                 {
                     return true;
                 }
 
-                var implementationMethod = method.ContainingType.FindImplementationForInterfaceMember(constructedMethod);
-                return method.Equals(implementationMethod);
+                // check whether 'method', when called on this receiver, is an implementation of 'constructedMethod'.
+                for (var baseType = receiverType; baseType is object; baseType = baseType.BaseTypeNoUseSiteDiagnostics)
```

Comment:
>baseType is object [](start = 50, length = 18)

For the negative case, we always traverse the entire inheritance hierarchy. We also always check all overridden methods, even those that are declared "below" ```baseType```, those will never be returned by ```FindImplementationForInterfaceMember```. I think this is an overkill.

Reply:
Would this approach work?
```
for (var baseType = receiverType; baseType is object;)
{
    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
    if (method.Equals(implementationMethod))
    {
        return true;
     }

    if (method.ContainingType.Equals(baseType))
    {
        method = method.OverriddenMethod
        baseType = method?.ContainingType;
    }
    else
    {
        baseType = baseType.BaseTypeNoUseSiteDiagnostics
    }
}
```

---
In reply to: [384828048](https://github.com/dotnet/roslyn/pull/41918#discussion_r384828048) [](ancestors = 384828048)

Reply:
Well, doing ```baseType = method?.ContainingType;``` is wrong. We should do ```baseType = baseType.BaseTypeNoUseSiteDiagnostics``` for both cases, but ```break``` once ```method``` is null.

---
In reply to: [384830098](https://github.com/dotnet/roslyn/pull/41918#discussion_r384830098) [](ancestors = 384830098,384828048)

Reply:
Also, to handle some cases that can only arise from metadata we probably should put the loop through overridden methods back. Here is the new revision:
```
for (var baseType = receiverType; baseType is object && method is object; baseType = baseType.BaseTypeNoUseSiteDiagnostics)
{
    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
    if (implementationMethod is object)
    {
        for (var overriddenMethod = method; overriddenMethod is object; overriddenMethod = overriddenMethod.OverriddenMethod)
        {
            if (overriddenMethod.Equals(implementationMethod))
            {
                return true;
            }
        }
    }

    if (method.ContainingType.Equals(baseType))
    {
        method = method.OverriddenMethod
    }
}
```


---
In reply to: [384834692](https://github.com/dotnet/roslyn/pull/41918#discussion_r384834692) [](ancestors = 384834692,384830098,384828048)

Reply:
I think I have done part of your suggestion by bailing out after checking a `baseType` when it is the same as the `method.ContainingType`.

For the other part: does a method being "below" the `baseType` mean that the method is declared in a derived type? Is this really a distinct optimization or is it a different way of saying the same thing as the first optimization? Maybe we can sync offline tomorrow?